api/segments/models.py

```diff
@@ -198,6 +198,93 @@ def deep_clone(self) -> "Segment":
 
         return cloned_segment
 
+    def assign_matching_rules_to_segment(self, segment: "Segment") -> bool:
+        """
+        Assign matching rules of the current object to the rules of the given segment.
+
+        This method iterates through the rules of the provided `Segment` and
+        attempts to match them to the current object's rules and sub-rules,
+        updating versioning where matches are found.
+
+        This is done in order to set the `version_of` field for matched rules
+        and conditions so that the frontend can more reliably diff rules and
+        conditions between a change request for a segment and the target segment
+        itself.
+
+        Args:
+            segment (Segment): The segment whose rules are being matched
+                               against the current object's rules.
+
+        Returns:
+            bool:
+                - `True` if any rules or sub-rules match between the current
+                  object and the segment.
+                - `False` if no matches are found.
+
+        Process:
+            1. Retrieve all rules associated with the current object and the segment.
+            2. For each rule in the segment:
+               - Check its sub-rules against the current object's rules and sub-rules.
+               - A match is determined if the sub-rule's type and properties align
+                 with those of the current object's sub-rules.
+               - If a match is found:
+                 - Update the `version_of` field for the matched sub-rule and rule.
+                 - Track the matched rules and sub-rules to avoid duplicate processing.
+            3. Perform a bulk update on matched rules and sub-rules to persist
+               versioning changes.
+
+        Side Effects:
+            - Updates the `version_of` field for matched rules and sub-rules.
+        """
+
+        self_rules = self.rules.all()
+        matched_rules = set()
+        matched_sub_rules = set()
+
+        for rule in segment.rules.all():
+            for sub_rule in rule.rules.all():
+
+                # Set a bool flag to be used to break out of the
+                # below self_rules iteration if a sub_rule matches.
+                sub_rule_matched = False
+                for self_rule in self_rules:
+                    if sub_rule_matched:
+                        break
+
+                    # If a self_rule has already been matched then
+                    # the only time we care about it is when we
+                    # further iterate on the the same sub_rules
+                    # that are present in the matching parent rule.
```

Comment:
Again, why? This logic is super hard to follow and all these comments are doing is just rephrasing the code that I can read already. It's super hard to follow what self_rule, sub_rule and 'parent rule' here are. Can we reword this to talk about the business logic rather than the code?

Reply:
Ok I've tried to do this one as well.